File: typescript/src/transport/webrtc/WebRtcTransport.ts
/**
 * Direct peer-to-peer transport for AetherNet over a WebRTC data channel. The WebRTC backend is
 * pluggable behind the {@link PeerLink} seam: Node uses werift (`peer-link.werift.ts`), the browser
 * uses the native `RTCPeerConnection` (`peer-link.dom.ts`). This class holds only the transport
 * logic — peer bookkeeping and signalling routing — so the identical code runs on both platforms.
 *
 * Serverless by default: with the default (no ICE servers) a node never contacts a STUN/TURN
 * server — host-candidate-only ICE forms a direct link on the same LAN or when a peer has a public
 * address. STUN/TURN are OPTIONAL (opted into by passing an explicit ICE-server list) and help
 * traverse NATs that host candidates alone can't. The SDP/ICE handshake (trickled) rides an
 * injected {@link Signaling} channel, so no central signalling server is required either. Implements
 * {@link ITransportService} so the transport ladder can rank it between the radio mesh (cheap,
 * proximity) and the relay (last resort): a direct internet path is used when one can be negotiated.
 *
 * Mirrors the C# `WebRtcTransportService` (SIPSorcery) and the Go `WebRtcTransport` (pion).
 *
 * SPDX-License-Identifier: MIT
 */

import { ITransportService, PerTransportMetrics } from "../ITransportService.js";
import {
  getDefaultPeerLinkFactory,
  type IceServer,
  type PeerLink,
  type PeerLinkFactory,
} from "./peer-link.js";
import { Signal, Signaling, SignalType } from "./signaling.js";

/** How long {@link WebRtcTransport.sendAsync} waits for the data channel to open. */
const CONNECT_TIMEOUT_MS = 20_000;

/**
 * Serverless default: NO ICE servers — a peer using the default never contacts a STUN/TURN
 * server. Direct links form on the same LAN or when a peer has a public address; for NAT
 * traversal without a server, route through the circuit-relay-v2 transport (peers relay for
 * peers). Opt into STUN/TURN by passing an explicit list, e.g.
 * `[{ urls: "stun:stun.l.google.com:19302" }]`.
 */
export function defaultIceServers(): IceServer[] {
  return [];
}

/**
 * Direct P2P transport over a WebRTC `RTCDataChannel`, backend-agnostic. The backend is werift on
 * Node and the native `RTCPeerConnection` in the browser — selected by the registered default (the
 * entry point you import) or an explicit `peerLinkFactory` constructor argument.
 *
 * With `undefined` `iceServers` the transport uses the serverless default of NO ICE servers
 * (host-candidate-only ICE) — it never contacts a STUN/TURN server, and links form on the same LAN
 * or when a peer has a public address. For NAT traversal without a server, route through the
 * circuit-relay-v2 transport. Pass an explicit list to opt into STUN/TURN.
 */
export class WebRtcTransport implements ITransportService {
  private readonly localUhid: string;
  private readonly signaling: Signaling;
  private readonly iceServers: IceServer[];
  private readonly peerLinkFactory?: PeerLinkFactory;
  private readonly peers = new Map<string, PeerLink>();
  private disposed = false;

  readonly name: string = "WebRTC P2P";
  maxBandwidthBps: number = 100_000_000; // direct link — bounded by the local NIC
  maxRangeMeters: number = 0;            // internet — unbounded
  powerCostRelative: number = 5;         // dearer than local radio on the 1-10 scale
  maxConcurrentPeers: number = 256;
  readonly metrics: PerTransportMetrics = new PerTransportMetrics();
  onDataReceived?: (senderUhid: string, data: Uint8Array) => void;

  /**
   * @param localUhid  This node's UHID.
   * @param signaling  The channel carrying SDP/ICE between peers.
   * @param iceServers `undefined` => the serverless default (NO ICE servers; host-candidate-only,
   *                   no STUN/TURN). An explicit list is respected verbatim.
   * @param peerLinkFactory Optional WebRTC backend override. Defaults to the registered backend
   *                   (werift on Node, native DOM in the browser).
   */
  constructor(
    localUhid: string,
    signaling: Signaling,
    iceServers?: IceServer[],
    peerLinkFactory?: PeerLinkFactory,
  ) {
    if (!localUhid || localUhid.trim().length === 0) {
      throw new Error("WebRtcTransport: localUhid must not be empty");
    }
    if (!signaling) {
      throw new Error("WebRtcTransport: signaling is required");
    }
    this.localUhid = localUhid;
    this.signaling = signaling;
    this.iceServers = iceServers ?? defaultIceServers();
    this.peerLinkFactory = peerLinkFactory;
    this.signaling.onSignal((signal) => {
      void this.handleSignal(signal);
    });
  }

  get isAvailable(): boolean {
    return !this.disposed;
  }

  // ── ITransportService ───────────────────────────────────────────────────────

  async sendAsync(
    peerUhid: string,
    data: Uint8Array,
    _cancellationToken?: AbortSignal,
  ): Promise<boolean> {
    if (this.disposed || !peerUhid || peerUhid.trim().length === 0) return false;
    if (data.length === 0) return false;

    const link = await this.getOrCreateLink(peerUhid, true);
    if (link === undefined) return false;

    const start = Date.now();
    const ok = await link.send(data, CONNECT_TIMEOUT_MS);
    const elapsed = Date.now() - start;
    this.metrics.recordSample(elapsed, ok, ok ? data.length : 0);
    return ok;
  }

  async sendStreamAsync(
    peerUhid: string,
    stream: ReadableStream<Uint8Array>,
    cancellationToken?: AbortSignal,
  ): Promise<boolean> {
    if (this.disposed) return false;

    const chunks: Uint8Array[] = [];
    const reader = stream.getReader();
    try {
      for (;;) {
        const { done, value } = await reader.read();
        if (done) break;
        if (value) chunks.push(value);
      }
    } finally {
      reader.releaseLock();
    }

    const total = chunks.reduce((sum, c) => sum + c.length, 0);
    const combined = new Uint8Array(total);
    let offset = 0;
    for (const chunk of chunks) {
      combined.set(chunk, offset);
      offset += chunk.length;
    }
    return this.sendAsync(peerUhid, combined, cancellationToken);
  }

  isConnected(peerUhid: string): boolean {
    if (this.disposed || !peerUhid) return false;
    const link = this.peers.get(peerUhid);
    return link !== undefined && link.isOpen;
  }

  /** Tears down all peer connections. */
  async dispose(): Promise<void> {
    if (this.disposed) return;
    this.disposed = true;
    const links = [...this.peers.values()];
    this.peers.clear();
    for (const link of links) {
      await link.close();
    }
    this.onDataReceived = undefined;
  }

  // ── Signalling inbound ──────────────────────────────────────────────────────

  private async handleSignal(signal: Signal): Promise<void> {
    if (this.disposed || signal.toUhid !== this.localUhid) return;
    try {
      switch (signal.type) {
        case SignalType.Offer: {
          const link = await this.getOrCreateLink(signal.fromUhid, false);
          if (link !== undefined && signal.sdp !== undefined) {
            await link.acceptOffer(signal.sdp);
          }
          break;
        }
        case SignalType.Answer: {
          const link = this.peers.get(signal.fromUhid);
          if (link !== undefined && signal.sdp !== undefined) {
            await link.acceptAnswer(signal.sdp);
          }
          break;
        }
        case SignalType.Candidate: {
          const link = this.peers.get(signal.fromUhid);
          if (link !== undefined) {
            await link.addRemoteCandidate(signal);
          }
          break;
        }
      }
    } catch {
      // A signalling failure must not crash the transport; ICE re-gathers on reconnect.
    }
  }

  private async getOrCreateLink(
    peerUhid: string,
    asInitiator: boolean,
  ): Promise<PeerLink | undefined> {
    if (this.disposed) return undefined;

    const existing = this.peers.get(peerUhid);
    if (existing !== undefined && !existing.isClosed) {
      if (asInitiator) await existing.waitOpen(CONNECT_TIMEOUT_MS);
      return existing;
    }

    const factory = this.peerLinkFactory ?? getDefaultPeerLinkFactory();
    const link = factory({
      localUhid: this.localUhid,
      peerUhid,
      iceServers: this.iceServers,
      signaling: this.signaling,
      onData: (from, bytes) => this.onPeerData(from, bytes),
    });
    this.peers.set(peerUhid, link);
    link.onClosed = () => {
      if (this.peers.get(peerUhid) === link) this.peers.delete(peerUhid);
    };

    await link.start(asInitiator);
    if (asInitiator) await link.waitOpen(CONNECT_TIMEOUT_MS);
    return link;
  }

  private onPeerData(peerUhid: string, data: Uint8Array): void {
    this.onDataReceived?.(peerUhid, data);
  }
}

File: kotlin/src/main/kotlin/aethernet/transport/webrtc/WebRtcTransport.kt
// SPDX-License-Identifier: MIT

package aethernet.transport.webrtc

import aethernet.transport.PerTransportMetrics
import aethernet.transport.TransportService
import dev.onvoid.webrtc.CreateSessionDescriptionObserver
import dev.onvoid.webrtc.PeerConnectionFactory
import dev.onvoid.webrtc.PeerConnectionObserver
import dev.onvoid.webrtc.RTCAnswerOptions
import dev.onvoid.webrtc.RTCConfiguration
import dev.onvoid.webrtc.RTCDataChannel
import dev.onvoid.webrtc.RTCDataChannelBuffer
import dev.onvoid.webrtc.RTCDataChannelInit
import dev.onvoid.webrtc.RTCDataChannelObserver
import dev.onvoid.webrtc.RTCDataChannelState
import dev.onvoid.webrtc.RTCIceCandidate
import dev.onvoid.webrtc.RTCIceServer
import dev.onvoid.webrtc.RTCOfferOptions
import dev.onvoid.webrtc.RTCPeerConnection
import dev.onvoid.webrtc.RTCPeerConnectionState
import dev.onvoid.webrtc.RTCRtpReceiver
import dev.onvoid.webrtc.RTCRtpTransceiver
import dev.onvoid.webrtc.RTCSdpType
import dev.onvoid.webrtc.RTCSessionDescription
import dev.onvoid.webrtc.SetSessionDescriptionObserver
import dev.onvoid.webrtc.media.MediaStream
import dev.onvoid.webrtc.media.audio.AudioDeviceModule
import dev.onvoid.webrtc.media.audio.AudioLayer
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.withTimeoutOrNull
import org.slf4j.LoggerFactory
import java.nio.ByteBuffer
import java.util.concurrent.ConcurrentHashMap

/**
 * Direct peer-to-peer transport over a WebRTC data channel
 * ([dev.onvoid.webrtc], a JVM wrapper around Google's native libwebrtc).
 *
 * NAT traversal is handled by ICE/STUN, with WebRTC's own TURN as last resort. The initial SDP/ICE
 * handshake is carried by an injected [WebRtcSignaling] channel (e.g. the AetherNet relay), so no
 * central signalling server is required. Implements [TransportService] so the transport ranker can
 * place it between the radio mesh (cheap, proximity) and the QUIC/HTTP relay (last resort) — a
 * direct internet path is used when one can be negotiated, otherwise the relay carries the traffic.
 *
 * This is the Kotlin/JVM sibling of the SIPSorcery-backed C# transport and the pion-backed Go
 * transport, giving the JVM core its first real, internet-capable transport (the others are
 * in-process simulations).
 *
 * Received bytes are surfaced on [dataReceived] as `(senderUhid, data)` pairs — the same
 * [kotlinx.coroutines.flow.Flow] idiom `InProcessTransport` uses.
 *
 * @param localUhid  This node's UHID.
 * @param signaling  The signalling channel carrying SDP/ICE to and from peers.
 * @param iceServers `null` uses the serverless default of NO ICE servers (host-candidate-only ICE)
 *                   — it never contacts a STUN/TURN server, and links form on the same LAN or when
 *                   a peer has a public address. For NAT traversal without a server, route through
 *                   the circuit-relay-v2 transport (peers relay for peers). Pass an explicit list to
 *                   opt into STUN/TURN; an explicit empty list keeps host-candidate-only ICE.
 */
class WebRtcTransport(
    private val localUhid: String,
    private val signaling: WebRtcSignaling,
    iceServers: List<RTCIceServer>? = null,
) : TransportService {

    init {
        require(localUhid.isNotEmpty()) { "localUhid required" }
    }

    private val log = LoggerFactory.getLogger(WebRtcTransport::class.java)

    private val iceServers: List<RTCIceServer> = iceServers ?: defaultIceServers()

    // The platform's WebRTC entry point. Native + heavyweight, so one per transport, lazily built
    // and disposed on close.
    //
    // Built with a DUMMY AudioDeviceModule. This is a data-channel-only transport that never
    // carries audio, but the default `PeerConnectionFactory()` makes libwebrtc bring up real audio
    // hardware and hard-abort ("Failed to initialize the ADM") on any host without a usable sound
    // card — a headless Circle OS gateway/relay node, a server, or a CI runner. `kDummyAudio` needs
    // no device, so the data path works everywhere. The ADM is held so it can be disposed on close.
    private var audioModule: AudioDeviceModule? = null
    private val factory: PeerConnectionFactory by lazy {
        val adm = AudioDeviceModule(AudioLayer.kDummyAudio)
        audioModule = adm
        PeerConnectionFactory(adm)
    }

    private val peers = ConcurrentHashMap<String, PeerLink>()

    @Volatile
    private var closed = false

    private val mutableDataReceived = MutableSharedFlow<Pair<String, ByteArray>>(
        replay = 0,
        extraBufferCapacity = 256,
    )

    init {
        signaling.onSignal(::onSignal)
    }

    // ── TransportService ──────────────────────────────────────────────────────

    override val name: String = "WebRTC P2P"
    override val isAvailable: Boolean get() = !closed
    override val maxBandwidthBps: Long = 100_000_000L   // direct link — bounded by the local NIC
    override val maxRangeMeters: Int = 0                 // internet — unbounded
    override val powerCostRelative: Int = 5              // dearer than local radio on the 1-10 scale
    override val maxConcurrentPeers: Int = 256

    /** Non-null metrics instance; always present for this transport. */
    override val metrics: PerTransportMetrics = PerTransportMetrics()

    override val dataReceived: Flow<Pair<String, ByteArray>> = mutableDataReceived.asSharedFlow()

    override suspend fun sendAsync(peerUhid: String, data: ByteArray): Boolean {
        if (closed || peerUhid.isEmpty()) return false

        val link = getOrCreateLink(peerUhid, asInitiator = true) ?: return false
        val startMs = System.currentTimeMillis()
        val ok = link.send(data)
        val rttMs = (System.currentTimeMillis() - startMs).coerceAtLeast(1L)
        metrics.recordSample(rttMs, success = ok, bytesTransferred = if (ok) data.size.toLong() else 0L)
        return ok
    }

    override suspend fun sendStreamAsync(peerUhid: String, data: ByteArray): Boolean =
        sendAsync(peerUhid, data)

    override fun isConnected(peerUhid: String): Boolean =
        peers[peerUhid]?.isOpen == true

    override fun close() {
        if (closed) return
        closed = true
        for (link in peers.values) {
            link.close()
        }
        peers.clear()
        try {
            factory.dispose()
        } catch (e: Throwable) {
            log.warn("[WebRTC] factory dispose failed", e)
        }
        try {
            audioModule?.dispose()
        } catch (e: Throwable) {
            log.warn("[WebRTC] audio module dispose failed", e)
        }
    }

    // ── Signalling inbound ──────────────────────────────────────────────────────

    private fun onSignal(signal: WebRtcSignal) {
        if (closed || signal.toUhid != localUhid) return
        try {
            when (signal.type) {
                WebRtcSignalType.OFFER -> {
                    val link = getOrCreateLink(signal.fromUhid, asInitiator = false)
                    val sdp = signal.sdp
                    if (link != null && sdp != null) link.acceptOffer(sdp)
                }

                WebRtcSignalType.ANSWER -> {
                    val sdp = signal.sdp
                    if (sdp != null) peers[signal.fromUhid]?.acceptAnswer(sdp)
                }

                WebRtcSignalType.ICE_CANDIDATE ->
                    peers[signal.fromUhid]?.addRemoteCandidate(signal)
            }
        } catch (e: Throwable) {
            log.warn("[WebRTC] signal handling failed from {}", signal.fromUhid, e)
        }
    }

    private fun getOrCreateLink(peerUhid: String, asInitiator: Boolean): PeerLink? {
        if (closed) return null

        val existing = peers[peerUhid]
        if (existing != null && !existing.isClosed) return existing

        val link = PeerLink(peerUhid)
        val winner = peers.putIfAbsent(peerUhid, link)
        if (winner != null) {
            // Lost a race — discard ours, use the winner.
            link.close()
            return winner
        }

        link.start(asInitiator)
        return link
    }

    private fun onPeerData(peerUhid: String, data: ByteArray) {
        // Native callback thread — emit without suspending. extraBufferCapacity makes tryEmit
        // succeed under normal load; if a slow/absent collector ever fills the buffer the oldest
        // pending datum is dropped (best-effort delivery, like the reference transports).
        if (!mutableDataReceived.tryEmit(peerUhid to data)) {
            log.warn("[WebRTC] dataReceived buffer full; dropped {} bytes from {}", data.size, peerUhid)
        }
    }

    // ── One WebRTC connection to a single peer ─────────────────────────────────

    private inner class PeerLink(private val peerUhid: String) : PeerConnectionObserver {

        private val pc: RTCPeerConnection
        private val open = CompletableDeferred<Boolean>()

        @Volatile
        private var channel: RTCDataChannel? = null

        @Volatile
        private var terminated = false

        val isOpen: Boolean get() = channel?.state == RTCDataChannelState.OPEN
        val isClosed: Boolean get() = terminated

        init {
            val config = RTCConfiguration().apply {
                this.iceServers.addAll(this@WebRtcTransport.iceServers)
            }
            pc = factory.createPeerConnection(config, this)
        }

        /** Begins the handshake. The initiator creates the data channel + sends the offer. */
        fun start(asInitiator: Boolean) {
            if (!asInitiator) return // responder waits for the inbound offer (see acceptOffer)

            val dc = pc.createDataChannel(DATA_CHANNEL_LABEL, RTCDataChannelInit())
            attachChannel(dc)

            pc.createOffer(RTCOfferOptions(), object : CreateSessionDescriptionObserver {
                override fun onSuccess(description: RTCSessionDescription) {
                    pc.setLocalDescription(description, object : SetSessionDescriptionObserver {
                        override fun onSuccess() {
                            signaling.sendSignal(
                                peerUhid,
                                WebRtcSignal(
                                    fromUhid = localUhid,
                                    toUhid = peerUhid,
                                    type = WebRtcSignalType.OFFER,
                                    sdp = description.sdp,
                                ),
                            )
                        }

                        override fun onFailure(error: String) {
                            log.warn("[WebRTC] setLocalDescription(offer) failed for {}: {}", peerUhid, error)
                        }
                    })
                }

                override fun onFailure(error: String) {
                    log.warn("[WebRTC] createOffer failed for {}: {}", peerUhid, error)
                }
            })
        }

        fun acceptOffer(sdp: String) {
            val remote = RTCSessionDescription(RTCSdpType.OFFER, sdp)
            pc.setRemoteDescription(remote, object : SetSessionDescriptionObserver {
                override fun onSuccess() {
                    pc.createAnswer(RTCAnswerOptions(), object : CreateSessionDescriptionObserver {
                        override fun onSuccess(description: RTCSessionDescription) {
                            pc.setLocalDescription(description, object : SetSessionDescriptionObserver {
                                override fun onSuccess() {
                                    signaling.sendSignal(
                                        peerUhid,
                                        WebRtcSignal(
                                            fromUhid = localUhid,
                                            toUhid = peerUhid,
                                            type = WebRtcSignalType.ANSWER,
                                            sdp = description.sdp,
                                        ),
                                    )
                                }

                                override fun onFailure(error: String) {
                                    log.warn("[WebRTC] setLocalDescription(answer) failed for {}: {}", peerUhid, error)
                                }
                            })
                        }

                        override fun onFailure(error: String) {
                            log.warn("[WebRTC] createAnswer failed for {}: {}", peerUhid, error)
                        }
                    })
                }

                override fun onFailure(error: String) {
                    log.warn("[WebRTC] setRemoteDescription(offer) failed for {}: {}", peerUhid, error)
                }
            })
        }

        fun acceptAnswer(sdp: String) {
            val remote = RTCSessionDescription(RTCSdpType.ANSWER, sdp)
            pc.setRemoteDescription(remote, object : SetSessionDescriptionObserver {
                override fun onSuccess() = Unit
                override fun onFailure(error: String) {
                    log.warn("[WebRTC] setRemoteDescription(answer) failed for {}: {}", peerUhid, error)
                }
            })
        }

        fun addRemoteCandidate(signal: WebRtcSignal) {
            val candidate = signal.candidate ?: return
            if (candidate.isEmpty()) return
            pc.addIceCandidate(RTCIceCandidate(signal.sdpMid, signal.sdpMLineIndex, candidate))
        }

        private fun attachChannel(dc: RTCDataChannel) {
            channel = dc
            dc.registerObserver(object : RTCDataChannelObserver {
                override fun onBufferedAmountChange(previousAmount: Long) = Unit

                override fun onStateChange() {
                    when (dc.state) {
                        RTCDataChannelState.OPEN -> open.complete(true)
                        RTCDataChannelState.CLOSED -> markClosed()
                        else -> Unit
                    }
                }

                override fun onMessage(buffer: RTCDataChannelBuffer) {
                    val data = buffer.data
                    val bytes = ByteArray(data.remaining())
                    data.get(bytes)
                    onPeerData(peerUhid, bytes)
                }
            })
            // The channel may already be OPEN by the time the observer is registered.
            if (dc.state == RTCDataChannelState.OPEN) open.complete(true)
        }

        // PeerConnectionObserver

        override fun onIceCandidate(candidate: RTCIceCandidate) {
            signaling.sendSignal(
                peerUhid,
                WebRtcSignal(
                    fromUhid = localUhid,
                    toUhid = peerUhid,
                    type = WebRtcSignalType.ICE_CANDIDATE,
                    candidate = candidate.sdp,
                    sdpMid = candidate.sdpMid,
                    sdpMLineIndex = candidate.sdpMLineIndex,
                ),
            )
        }

        override fun onDataChannel(dataChannel: RTCDataChannel) {
            // Responder receives the channel the initiator opened.
            attachChannel(dataChannel)
        }

        override fun onConnectionChange(state: RTCPeerConnectionState) {
            when (state) {
                RTCPeerConnectionState.FAILED,
                RTCPeerConnectionState.DISCONNECTED,
                RTCPeerConnectionState.CLOSED -> markClosed()
                else -> Unit
            }
        }

        override fun onRenegotiationNeeded() = Unit

        override fun onAddTrack(receiver: RTCRtpReceiver, mediaStreams: Array<out MediaStream>) = Unit

        override fun onRemoveTrack(receiver: RTCRtpReceiver) = Unit

        override fun onTrack(transceiver: RTCRtpTransceiver) = Unit

        // Link lifecycle

        private fun markClosed() {
            if (terminated) return
            terminated = true
            open.complete(false)
        }

        suspend fun waitOpen(timeoutMs: Long): Boolean {
            if (isOpen) return true
            if (terminated) return false
            return withTimeoutOrNull(timeoutMs) { open.await() } ?: false
        }

        suspend fun send(data: ByteArray): Boolean {
            if (!waitOpen(CONNECT_TIMEOUT_MS)) return false
            val dc = channel ?: return false
            return try {
                dc.send(RTCDataChannelBuffer(ByteBuffer.wrap(data), true))
                true
            } catch (e: Throwable) {
                log.warn("[WebRTC] send to {} failed", peerUhid, e)
                false
            }
        }

        fun close() {
            markClosed()
            try {
                channel?.close()
            } catch (_: Throwable) { /* best effort */ }
            try {
                channel?.dispose()
            } catch (_: Throwable) { /* best effort */ }
            try {
                pc.close()
            } catch (_: Throwable) { /* best effort */ }
        }
    }

    companion object {
        private const val DATA_CHANNEL_LABEL = "aether"
        private const val CONNECT_TIMEOUT_MS = 20_000L

        /**
         * Serverless default: NO ICE servers, so a node never contacts a STUN/TURN server. Direct
         * links form on the same LAN or when a peer has a public address; for NAT traversal without
         * a server, route through the circuit-relay-v2 transport (peers relay for peers). Callers
         * opt into STUN/TURN by passing an explicit list.
         */
        fun defaultIceServers(): List<RTCIceServer> = emptyList()
    }
}
